Rate and review

MicahSchiewe opened the pull request:
I'm pretty sure parts of it need help. Let me know what you see.

MicahSchiewe committed:
Failed attempt to genericize logic

* Determined that the logic did not work, would require extensive research
  into how Typescript's typing works to make it work right.
* That, and some API calls

MicahSchiewe committed:
Create initial backend API for saving/retrieving reviews

MicahSchiewe committed:
Add delete reviews

MicahSchiewe committed:
TMP

MicahSchiewe committed:
Re-arrange Review API to support high cohesion

* Tracking where the various Review endpoints were was hard, so I added a
  new class to hold all of them

MicahSchiewe committed:
Update API files to reflect backend changes

MicahSchiewe committed:
Add initial view review page

MicahSchiewe committed:
Begin adding create review page

* Impasse: the Form does not allow us to submit attributes which are not
  customizable. This is a problem which must be addressed.
* I accidentally saved a TMP commit, didn't I? Oops.

MicahSchiewe committed:
Add initial decorator for non-editable attributes in the form

MicahSchiewe committed:
Continue work on review form

* Start to make it render

MicahSchiewe committed:
Add create review interface

MicahSchiewe committed:
Clean up UI

MicahSchiewe committed:
Wire leaving reviews to view truck page

MicahSchiewe committed:
Make create review page properly load previous review

* Thus making it create + edit old
* No double-posting allowed! Not even the illusion!
* I need sleep...

MicahSchiewe committed:
Eclectic fix

* Fix hardcoded spacing between reviews
* Fix bug causing crashes with non-extended reviews

MicahSchiewe committed:
Add deleting reviews to frontend

* Also streamline review routing

MicahSchiewe committed:
TMP

MicahSchiewe committed:
Bigfix

* Make review able to be switched extended -> non-extended
* Minor undefined attribute access fix in Review.backendToFrontend

MicahSchiewe committed:
More bigfixes

* Remove unneccessary bookkeeping in RateReviewComponent
* Fix backend-to-frontend attribute mapping

MicahSchiewe committed:
Update Form to accomodate dynamic forms

* Now, if a field is dynamically removed, it is ACTUALLY removed from the
  form (instead of just languishing there and getting submitted anyway)

MicahSchiewe committed:
Fix losing changes on toggling extended <-> not

* Still loses changes to review, but retains the original

MicahSchiewe committed:
Add chop-shop review and star rating on truck

* I couldn't get the fancy Spring Interceptors or Hibernate Event
  listeners to work. Thus, this just calls a method at the end of Review
  service operations that update reviews. This is... REALLY BAD. I KNOW.
  But I don't have time to keep banging my head against this with Milestone
  III coming up. Sheesh, I didn't have the afternoons I ALREADY blew on
  this. I need a Dr. Pepper.

MicahSchiewe committed:
Merge branch 'master' of https://github.com/Spring-Kings/food-truck-finder into rate-and-review

# Conflicts:
#	food-truck-api/src/main/java/food/truck/api/endpoint/TruckEndpoint.java
#	food-truck-api/src/main/java/food/truck/api/endpoint/UserEndpoint.java
#	food-truck-frontend/src/components/TruckView.tsx
#	food-truck-frontend/src/pages/truck/[truckId].tsx
#	food-truck-frontend/src/pages/truck/edit/[truckId].tsx

MicahSchiewe committed:
Fix leave review button

* Don't offer to let a truck owner leave a review

nnazo reviewed (commented):
Looks great!

KevinDeMars commented on food-truck-api/src/main/java/food/truck/api/endpoint/ReviewEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+    public List<Review> getUserReviews(@RequestParam String username) {
+        User user = userService.loadUserByUsername(username);
+        if(user == null){
+            return new LinkedList<Review>();
```
Could return a 404 error

MicahSchiewe replied:
Probably a better solution. Will make this change.

KevinDeMars commented on food-truck-api/src/main/java/food/truck/api/endpoint/ReviewEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+    @Secured({"ROLE_USER"})
+    @PostMapping("/reviews/truck/{truckId}")
+    public Review postTruckReview(@AuthenticationPrincipal User user, @PathVariable long truckId, @RequestBody PostReviewParams data) {
+        // TODO decide if we want to just silently change the review to match
+        if (!user.getId().equals(data.userId))
+            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "UserID does not match userID of the provided review");
```
Why not just use the user's ID and leave out the userId param?

MicahSchiewe replied:
Will make that change. Amusingly enough, this removes the need for the HiddenAttributeForm. I'll leave it in in case we need it, though.

KevinDeMars commented on food-truck-api/src/main/java/food/truck/api/endpoint/ReviewEndpoint.java:
```diff
@@ -0,0 +1,94 @@
+    @Secured({"ROLE_USER"})
+    @DeleteMapping("/reviews/truck/{truckId}")
+    public boolean deleteReview(@AuthenticationPrincipal User u, @PathVariable long truckId) {
+        return reviewService.deleteReviewByUser(u, truckId);
```
These should be checked that the review belongs to the user

MicahSchiewe replied:
Authentication; what a concept! Will fix.

MicahSchiewe replied:
Forgot to follow up on this: so the way it fetches the review in the service, it actually doesn't need/can't be authenticated. It fetches the review matching the userId from AuthenticationPrinciple and the provided truckId, so by definition the user can't delete someone else's review unless they're signed in as someone else or defeat our authentication system, neither of which we can't check for.

KevinDeMars commented on food-truck-api/src/main/java/food/truck/api/reviews_and_subscriptions/ReviewService.java:
```diff
@@ -1,26 +1,92 @@
+        return reviewRepository.findByUserIdAndTruckId(userId, truckId).stream().findFirst().orElse(null);
     }
 
+    /**
+     * I hate this solution, but it didn't work out any other way...
+     */
```
Is this because you don't want to look at every single review to update average? I think it would work if you do a weighted average, like newAvg = (oldAvg * numReviews + newReview) / (numReviews + 1). But it would probably be less error-prone to just recalculate it.

MicahSchiewe replied:
The reason was because I know that this means we need to keep track of anywhere we want to update reviews and run this code there, which is asking for a bug (where we update reviews and forget to run it). I wanted to set it up as a trigger-ish thing on updating the Reviews table, but Hibernate did not want to cooperate.

KevinDeMars replied:
This sounds like a good way to me, only allowing changes to the reviews through the public methods in ReviewService, and updating the average in those methods

KevinDeMars commented on food-truck-frontend/src/api/RateReview.ts:
```diff
@@ -0,0 +1,201 @@
+    });
+
+    // If we got a good response, then create frontend representation
+    if (resp.data !== null) {
+      // Map all backend reviews to frontend reviews
+      // Learned to use `Promise.all` from: https://stackoverflow.com/questions/40140149/use-async-await-with-array-map
+      reviews.reviews = await Promise.all(resp.data.map(async (r: any) => {
+        // Fetch username
+        let name: any = await api.request({
+          url: `/user/${r.userId}`,
+          method: "GET",
+        });
```
It probably would have been better if the backend returned a "ReviewView" or something that had all the info we needed on the frontend. But since we already have the work in the frontend, might as well keep it.

KevinDeMars commented on food-truck-frontend/src/components/Form.tsx:
```diff
@@ -103,6 +103,17 @@ class Form extends Component<Props, State> {
         }));
     }
+
+    static getDerivedStateFromProps(props: Props, state: State) {
+        let newFormData: any = {};
+        if (props.children)
+            React.Children.forEach(props.children, (c: any) => {
+                let name: string = (c as ReactElement).props.name as string;
+                newFormData[name] = state.formData[name];
+            });
+        state.formData = newFormData;
+        return state;
+    }
```
What is this for?

MicahSchiewe replied:
This is to solve an issue with the Form not wanting to dynamically alter its content. The way React optimizes, it doesn't re-instantiate components when props change, it just re-runs the lifecycle methods. Since the state was formalized in the constructor, this meant that, should a form entry component be removed (for me, toggling extended to non-extended review dropping the text field for entering the review), its data would still exist in the form and be submitted (thus making it impossible to submit a non-extended review after saving an extended one).

As for what the method does, this is a life cycle method that runs when the props are updated, allowing you to update state based on it. While there are MANY anti-patterns revolving around this method, our use case (filtering out any state attributes that no longer existed in the form) fell under one of the few legitimate uses.

I'll confess there may still be a bug or two with it (I'm pretty sure adding a new component worked fine, since I could submit extended reviews in the first place when it defaulted to non-extended, but I can't swear to it). But I probably didn't test it enough.

KevinDeMars commented on food-truck-frontend/src/components/util/HiddenAttributeForm.tsx:
```diff
@@ -0,0 +1,34 @@
+import { TextField } from "@material-ui/core";
+import React from "react";
+import Form, { Props as FormProps } from "../Form";
+
+export interface HiddenEntry {
+  name: string;
+  defaultValue: any;
+}
+
+export interface HiddenAttributeFormProps {
+  hiddenAttrs: HiddenEntry[];
+}
+
+class HiddenField extends React.Component<HiddenEntry> {
+  constructor(props: HiddenEntry) {
+    super(props);
+  }
+  render() {
+    return <div />;
+  }
+}
+
+function HiddenAttributeForm(props: HiddenAttributeFormProps & FormProps) {
+  return (
+    <Form {...props}>
+      {React.Children.map(props.children, c => c)}
+      {props.hiddenAttrs.map((tpl) => (
+        <HiddenField key={`hidden${tpl.name}`} {...tpl} />
+      ))}
+    </Form>
+  );
+}
+
+export default HiddenAttributeForm;
```
This looks like it will work, but we could also just use CSS "display: none"

MicahSchiewe replied:
Huh. Learn something new every day. Yeah, I could've used that. Well, it's working as-is, so I won't change it unless we decide that CSS property would work better.

MicahSchiewe committed:
Fix not loading API key

* A bit of a hack. Loads it from the environment variables of the machine,
  in stead of from the .env file. Since the .env file doesn't want to
  process right for us.

MicahSchiewe committed:
Update in response to suggestions

* Also, squash bugs

MicahSchiewe committed:
Merge branch 'master' of https://github.com/Spring-Kings/food-truck-finder into rate-and-review

# Conflicts:
#	food-truck-api/src/main/java/food/truck/api/endpoint/TruckEndpoint.java
#	food-truck-api/src/main/java/food/truck/api/reviews_and_subscriptions/ReviewService.java

nnazo committed:
fix material-ui/lab types

KevinDeMars commented on food-truck-api/src/main/java/food/truck/api/endpoint/ReviewEndpoint.java:
```diff
@@ -73,22 +76,26 @@ public Review getTruckReviews(@PathVariable long truckId, @RequestParam Long use
     @DeleteMapping("/reviews/truck/{truckId}")
     public boolean deleteReview(@AuthenticationPrincipal User u, @PathVariable long truckId) {
-        return reviewService.deleteReviewByUser(u, truckId);
+        boolean result = reviewService.deleteReviewByUser(u, truckId);
```
This should also be checked if the user owns the review

MicahSchiewe replied:
Accidentally posted this on the wrong iteration of this comment:
"Forgot to follow up on this: so the way it fetches the review in the service, it actually doesn't need/can't be authenticated. It fetches the review matching the userId from AuthenticationPrinciple and the provided truckId, so by definition the user can't delete someone else's review unless they're signed in as someone else or defeat our authentication system, neither of which we can check for."

KevinDeMars commented on food-truck-api/src/main/java/food/truck/api/reviews_and_subscriptions/ReviewService.java:
```diff
@@ -55,34 +57,34 @@ public Review saveReview(Long userId, Truck t, int score, int costRating, String
+            cost = coll.stream()
+                    .collect(Collectors.averagingInt(Review::getCostRating))
+                    .longValue();
+            star = coll.stream()
+                    .collect(Collectors.averagingInt(Review::getStarRating))
+                    .longValue();
```
For the average rating, I think it would make sense for it to be a double so a 4.95-star truck is treated better than a 4.01-star

MicahSchiewe replied:
I can look into it, just not sure how many hoops I'd need to jump through with the rating selector components on the frontend.

MicahSchiewe replied:
Added it

MicahSchiewe committed:
Merge branch 'master' of https://github.com/Spring-Kings/food-truck-finder into rate-and-review

MicahSchiewe committed:
Make average rating a floating point value

* So 4.1 and 4.9 are not treated the same
* Currently has just 0.1 precision, so 4.95 and 4.9 are considered the same

MicahSchiewe committed:
Fix backend truck endpoint test to accomodate shift to doubles

MicahSchiewe committed:
Fix endpoint update restrictions for price rating.

MicahSchiewe committed:
Merge branch 'master' of https://github.com/Spring-Kings/food-truck-finder into rate-and-review

AlejandroNavarro1 reviewed (approved):
Don't see anything I would fix, so its good for me
